### PR TITLE
Fix survey and thread devas 1819 devas 1820

### PR DESCRIPTION
**Pull request by cedricmessiant**
- Fix DEVAS-1819
- Fix DEVAS-1820
- Fix another bug with completed phases in question page
- Refactoring: rename `getIfPhaseCompletedById` in `getIsPhaseCompletedById`, add a `findPhase` function, use find instead of forEach
- throw errors when something is wrong in some of the timeline utils
- add unit tests for `getIsPhaseCompletedById`
- Refactoring: use `Props` and `State` for types

### CODE REVIEW

File: assembl/static2/js/app/components/debate/survey/navigation.jsx [hunk 1]
```diff
@@ -74,8 +73,7 @@ class Navigation extends React.Component {
     const limitToHide = getDomElementOffset(firstTextarea).top + firstTextarea.clientHeight;
     const limitToShow = limitToHide + document.getElementById('nav').clientHeight;
     const windowOffset = window.pageYOffset + this.props.screenHeight;
-    const { timeline, phaseId } = this.props;
-    const isPhaseCompleted = getIfPhaseCompletedById(timeline, phaseId);
+    const { isPhaseCompleted } = this.props;
```
Comment (amensouissi) [inline assembl/static2/js/app/components/debate/survey/navigation.jsx]: remove the timeline dep

File: assembl/static2/js/app/components/debate/survey/post.jsx [hunk 2]
```diff
@@ -61,11 +59,9 @@ class Post extends React.Component<Props> {
 
   handleSentiment = (event, type, refetchQueries, currentCounts: { disagree: number, like: number }) => {
     const { post } = this.props.data;
-    const { phaseId } = this.props;
+    const { isPhaseCompleted } = this.props;
     const isUserConnected = getConnectedUserId() !== null;
     if (isUserConnected) {
-      const { timeline } = this.props;
```
Comment (amensouissi) [inline assembl/static2/js/app/components/debate/survey/post.jsx]: and here

File: assembl/static2/js/app/components/debate/survey/question.jsx [hunk 3]
```diff
@@ -94,16 +92,14 @@ class Question extends React.Component<QuestionProps, QuestionState> {
...
-    const { phaseId, index, title, screenWidth, screenHeight, questionsLength } = this.props;
+    const { index, isPhaseCompleted, title, screenWidth, screenHeight, questionsLength } = this.props;
     const questionTitle = questionsLength > 1 ? `${index}/ ${title}` : title;
     let height = screenHeight;
     const timelineElm = document && document.getElementById('timeline');
     // This is necessary to bypass an issue with Flow
     if (timelineElm) {
       height = screenHeight - timelineElm.clientHeight;
     }
-    const { timeline } = this.props;
```
Comment (amensouissi) [inline assembl/static2/js/app/components/debate/survey/question.jsx]: and here :)

File: assembl/static2/js/app/components/debate/survey/posts.jsx [hunk 4]
```diff
@@ -74,6 +75,7 @@ export class DumbPosts extends React.Component<PostsProps> {
         loadPreviousMessage="debate.survey.loadRecentPosts"
         itemData={item => ({
           id: item.node.id,
+          isPhaseCompleted: isPhaseCompleted,
```
Comment (amensouissi) [inline assembl/static2/js/app/components/debate/survey/posts.jsx]: please remove phaseId (it's not used in the Post component)

File: assembl/static2/js/app/components/debate/survey/posts.jsx [hunk 5]
```diff
@@ -25,6 +25,7 @@ type PostsProps = {
   },
   networkStatus: number,
   defaultContentLocaleMapping: Map,
+  isPhaseCompleted: boolean,
```
Comment (amensouissi) [inline assembl/static2/js/app/components/debate/survey/posts.jsx]: remove phaseId

File: assembl/static2/js/app/pages/question.jsx [hunk 6]
```diff
@@ -61,7 +76,12 @@ export function DumbQuestion(props: QuestionProps) {
                 <h3 className="collapsed-title">
                   <span>{`${params.questionIndex}/ ${title}`}</span>
                 </h3>
-                <Posts questionId={params.questionId} themeId={thematicId} phaseId={phaseId} />
+                <Posts
+                  questionId={params.questionId}
+                  themeId={thematicId}
+                  phaseId={phaseId}
```
Comment (amensouissi) [inline assembl/static2/js/app/pages/question.jsx]: remove phaseId (not used)

File: assembl/static2/js/app/utils/timeline.js [hunk 7]
```diff
@@ -145,37 +145,33 @@ export const isCurrentPhase = (_phase: Phase) => {
...
-  const phaseId = _id;
-  let phaseName = '';
-  if (!phaseId || !timeline) return phaseName;
-  timeline.forEach((phase) => {
-    if (phase.id === phaseId) {
-      phaseName = phase.title;
-    }
-  });
-  return phaseName;
+function findPhase(timeline: Timeline, phaseId: string): Phase {
```
Comment (MilanBarande) [inline assembl/static2/js/app/utils/timeline.js]: Just curious: in what case do we favour the function keyword over the arrow syntax?

Comment (cedricmessiant) [reply assembl/static2/js/app/utils/timeline.js]: I favor the function keyword for functions that are on top level of a file. I think arrow functions are better for callbacks and functions defined at a deeper level (in a render for example). But it's a personal preference I guess.

**Pull request merged**